models/user.go

```diff
@@ -315,7 +317,14 @@ func (u *User) NewGitSig() *git.Signature {
 
 // EncodePasswd encodes password to safe format.
 func (u *User) EncodePasswd() {
-	newPasswd := base.PBKDF2([]byte(u.Passwd), []byte(u.Salt), 10000, 50, sha256.New)
+	var hashAlgorithm func() hash.Hash
+	switch setting.Cfg.Section("security").Key("PASSWORD_HASH_ALGORITHM").String() {
+	case "sha512":
+		hashAlgorithm = sha512.New
+	default:
```

Comment:
Also this should be done once on setup, not each time a user sets their password...